### PR TITLE
[ANCHOR-345] [ANCHOR-64] Implement the Event Processing Server

### PR DESCRIPTION
**Pull request by lijamie98**
<details>
  <summary>PR Checklist</summary>

### PR Structure

* [ ] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [ ] This PR's title starts with name of package that is most changed in the PR, ex.
  `paymentservice.stellar`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
</details>

### What

Implement the event processing server with the refactored `EventService` interface.

### Why
Feature implementation.

### Known limitations
The callback API and client status callback is done with dummy implementations. Will be added in the coming PRs.

### CODE REVIEW

File: core/src/main/java/org/stellar/anchor/event/EventService.java
Comment (philipliu) [inline core/src/main/java/org/stellar/anchor/event/EventService.java:76]: Does this map to a Kafka consumer ack, and DeleteMessageRequest in SQS?

File: core/src/main/java/org/stellar/anchor/event/EventService.java
Comment (philipliu) [inline core/src/main/java/org/stellar/anchor/event/EventService.java:89]: This is for SEP-24/31 transactions where as control is for callback?

File: core/src/main/java/org/stellar/anchor/event/EventService.java
Comment (philipliu) [reply core/src/main/java/org/stellar/anchor/event/EventService.java:76]: A SQS `ReceiveMessage` response can contain up to 10 messages. If any of those messages fail to process, we either ack all, or ack none. Should we have per message acks instead, or introduce a batch ack API?

Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/event/EventService.java:76]: > Does this map to a Kafka consumer ack, and DeleteMessageRequest in SQS?
I think this can be delegated to the implementation. In Kafka, it would be `commit` and in SQS, it would be `DeleteMessageRequest`. I estimate both Kafka and SQS will be able to take care of the `ack()` requirement.

> A SQS `ReceiveMessage` response can contain up to 10 messages. If any of those messages fail to process, we either ack all, or ack none. Should we have per message acks instead, or introduce a batch ack API?

This is a good point. However, per-message ack is not possible in Kafka. In this scenario, if the server is down, all messages will not be committed. If some of the messages cannot be processed, they will be sent to a DLQ. The rest will be processed. And all messages will be acked.

File: core/src/main/java/org/stellar/anchor/event/EventService.java
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/event/EventService.java:89]: No. The `CONTROL` is the place-holder for the [control-plane ](https://www.snaplogic.com/blog/data-plane-vs-control-plane-whats-the-difference#:~:text=The%20control%20plane%20decides%20how,instructions%20by%20forwarding%20the%20packets.). The comment is updated to

```java
  enum EventQueue {
    /** The event queue for the transaction events that are to be processed by the anchor. */
    TRANSACTION,

    /** The event queue for events that are used for control-plane purposes. */
    CONTROL
  }
```

Comment (philipliu) [reply core/src/main/java/org/stellar/anchor/event/EventService.java:89]: Are we using a queue for control plane related messages today?

Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/event/EventService.java:89]: No, it is not for related message today. It's just a hidden placeholder.

**Review by JiahuiWho — APPROVED** [review]

File: integration-tests/src/test/kotlin/org/stellar/anchor/platform/test/EventProcessingServerTests.kt
Comment (philipliu) [inline integration-tests/src/test/kotlin/org/stellar/anchor/platform/test/EventProcessingServerTests.kt:19]: Can this be replaced with `AnchorEvent` constructor? I don't think we need to test deserialization here.

File: platform/src/main/java/org/stellar/anchor/platform/component/sep/SepBeans.java
Comment (philipliu) [inline platform/src/main/java/org/stellar/anchor/platform/component/sep/SepBeans.java:58]: Nit: remove `println`

File: platform/src/main/java/org/stellar/anchor/platform/component/share/ClientsBeans.java
Comment (philipliu) [inline platform/src/main/java/org/stellar/anchor/platform/component/share/ClientsBeans.java:11]: Is this intentional? The default value is already `""`.

File: platform/src/main/java/org/stellar/anchor/platform/configurator/ConfigReader.java
Comment (philipliu) [inline platform/src/main/java/org/stellar/anchor/platform/configurator/ConfigReader.java:68]: Nit: if we compile this every time we match, we may as well not compile it.

File: platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java
Comment (philipliu) [inline platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java:76]: Should we be throwing an exception here?

File: platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java
Comment (philipliu) [inline platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java:28]: Typo: `callback-api-`

File: platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java
Comment (philipliu) [inline platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java:143]: Should we introduce some rate limiting when `getEvents` is 0? Maybe an exponential backoff?

File: platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java
Comment (philipliu) [inline platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java:137]: Remove `Kafka`

File: platform/src/main/java/org/stellar/anchor/platform/event/KafkaSession.java
Comment (philipliu) [inline platform/src/main/java/org/stellar/anchor/platform/event/KafkaSession.java:89]: Do we want to emit a metric for read as well?

File: core/src/main/java/org/stellar/anchor/event/EventService.java
Comment (philipliu) [inline core/src/main/java/org/stellar/anchor/event/EventService.java:18]: Nit: missing session name

File: platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java
Comment (philipliu) [inline platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java:103]: Would we need to initialize two queues for the SQS implementation, one for status callback and the other business server events?

**Review by philipliu — COMMENTED** [review]
Did a preliminary review. Will do another pass since this PR is quite large.

File: core/src/main/java/org/stellar/anchor/event/EventService.java
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/event/EventService.java:18]: Nice catch!

File: core/src/main/java/org/stellar/anchor/event/EventService.java
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/event/EventService.java:38]: I will leave it to the implementation of the `Session`s. For example, `KafkaSession`.

File: core/src/main/java/org/stellar/anchor/event/EventService.java
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/event/EventService.java:81]: The `AnchorEvent` already has a `enum Type` field. 

https://github.com/stellar/java-stellar-anchor-sdk/blob/develop/api-schema/src/main/java/org/stellar/anchor/api/event/AnchorEvent.java#L21

Is this something you are referring to?

File: core/src/main/java/org/stellar/anchor/sep24/Sep24Transaction.java
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/sep24/Sep24Transaction.java:294]: Although this is refactoring which is preferred to be in a separate PR, considering the pros and cons, I think we should do it in this PR. 

I had updated according to the suggestion.

File: core/src/main/java/org/stellar/anchor/sep38/Sep38Service.java
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/sep38/Sep38Service.java:43]: Ok.

File: integration-tests/src/test/kotlin/org/stellar/anchor/platform/test/EventProcessingServerTests.kt
Comment (lijamie98) [reply integration-tests/src/test/kotlin/org/stellar/anchor/platform/test/EventProcessingServerTests.kt:19]: We are not testing the deserialization. It is a convenient way to construct from JSON rather in Java. The pattern is used in AP.

File: platform/src/main/java/org/stellar/anchor/platform/component/sep/SepBeans.java
Comment (lijamie98) [reply platform/src/main/java/org/stellar/anchor/platform/component/sep/SepBeans.java:58]: Removed.

File: platform/src/main/java/org/stellar/anchor/platform/component/share/ClientsBeans.java
Comment (lijamie98) [reply platform/src/main/java/org/stellar/anchor/platform/component/share/ClientsBeans.java:11]: Sprint gives contradicting warnings. Removing the `""` gives another warning.

File: platform/src/main/java/org/stellar/anchor/platform/configurator/ConfigReader.java
Comment (lijamie98) [reply platform/src/main/java/org/stellar/anchor/platform/configurator/ConfigReader.java:68]: Optimized.

File: platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java
Comment (lijamie98) [reply platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java:28]: Don't know how you catch this, but awesome catch.

File: platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java
Comment (lijamie98) [reply platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java:76]: Can't really throw in lambda unless changing to `for` loop. Since this normally do not happen because it's an `enum` and relatively stable, the `errorF` should give strong warning as well.

Considering the pros and cons AGAIN, I will change to for-loop with throwing exceptions.

File: platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java
Comment (lijamie98) [reply platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java:103]: In SQS, I think we will need to have a pre-processor that dispatches the incoming queue to multiple queues. Then one processor for each queue.

File: platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java
Comment (lijamie98) [reply platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java:143]: The Kafka `read()` is blocking and has a `poll()` timeout. It is left to the implementing class.

File: platform/src/main/java/org/stellar/anchor/platform/event/KafkaSession.java
Comment (lijamie98) [reply platform/src/main/java/org/stellar/anchor/platform/event/KafkaSession.java:89]: I put a TODO here. It will be handled in [ANCHOR-371](https://stellarorg.atlassian.net/browse/ANCHOR-371).

File: core/src/main/java/org/stellar/anchor/event/EventService.java
Comment (philipliu) [reply core/src/main/java/org/stellar/anchor/event/EventService.java:81]: Right, so if the `Type` was `CONTROL` do the fields in `AnchorEvents` still make sense. I don't imagine `quote` will be relevant for control plane messages.

File: integration-tests/src/test/kotlin/org/stellar/anchor/platform/test/EventProcessingServerTests.kt
Comment (philipliu) [reply integration-tests/src/test/kotlin/org/stellar/anchor/platform/test/EventProcessingServerTests.kt:19]: I would prefer we if stop using this pattern because it's easier to see diffs if it was initialized using Java. We wouldn't need to worry about white space diffs in JSON text blocks.

This is just a nit, so we don't need to block the PR on this.

File: platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java
Comment (philipliu) [reply platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java:102]: Will this pre-processor be implemented in `EventService.Session`?

File: core/src/main/java/org/stellar/anchor/event/EventService.java
Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/event/EventService.java:81]: No, the `AnchorEvents` are for data plane messages only.

File: integration-tests/src/test/kotlin/org/stellar/anchor/platform/test/EventProcessingServerTests.kt
Comment (lijamie98) [reply integration-tests/src/test/kotlin/org/stellar/anchor/platform/test/EventProcessingServerTests.kt:19]: > it's easier to see diffs if it was initialized using Java

I am not sure if I got it correctly. The current approach deserialize JSON into Java object. The assert should diff on the Java objects.

File: platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java
Comment (lijamie98) [reply platform/src/main/java/org/stellar/anchor/platform/event/EventProcessorManager.java:102]: I am thinking starting the pre-processor in the bean construction.

**Pull request merged**
